l10n_th_personal_income_tax/models/pit_rate.py

```diff
@@ -0,0 +1,139 @@
+# Copyright 2021 Ecosoft Co., Ltd. (http://ecosoft.co.th)
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl).
+
+from odoo import _, api, fields, models
+from odoo.exceptions import UserError, ValidationError
+from odoo.tools.float_utils import float_compare
+
+from odoo.addons.l10n_th_withholding_tax_cert.models.withholding_tax_cert import (
+    WHT_CERT_INCOME_TYPE,
+)
+
+
+class PersonalIncomeTax(models.Model):
+    _name = "personal.income.tax"
+    _description = "PIT Table"
+    _rec_name = "calendar_year"
+
+    calendar_year = fields.Char(
+        required=True,
+        default=lambda self: fields.Date.context_today(self).strftime("%Y"),
+    )
+    effective_date = fields.Date(
+        string="Effective Date",
+        compute="_compute_effective_date",
+        store=True,
+    )
+    rate_ids = fields.One2many(
+        comodel_name="personal.income.tax.rate",
+        inverse_name="pit_id",
+        string="Withholding Tax Rates",
+    )
+    wt_account_id = fields.Many2one(
+        comodel_name="account.account",
+        string="Withholding Tax Account",
+        help="Default withholding tax account before payment",
+    )
+    wt_cert_income_type = fields.Selection(
+        WHT_CERT_INCOME_TYPE,
+        string="Type of Income",
+        help="Default withholding tax income type",
+    )
+    active = fields.Boolean(default=True)
+
+    @api.depends("calendar_year")
+    def _compute_effective_date(self):
+        for rec in self:
+            rec.effective_date = "{}-01-01".format(rec.calendar_year)
+
+    @api.constrains("rate_ids")
+    def _check_rate_ids(self):
+        for rec in self:
+            prev_income_to = 0.0
+            accum_tax = 0.0
+            for i, rate in enumerate(rec.rate_ids):
+                if i == 0 and rate.income_from != 0.0:
+                    raise UserError(_("Income amount must start from 0.0"))
+                if i > 0 and float_compare(rate.income_from, prev_income_to, 2) != 0:
+                    raise UserError(
+                        _(
+                            "Discontinued income range!\n"
+                            "Please make sure Income From = Previous Income To"
+                        )
+                    )
+                prev_income_to = rate.income_to
+                accum_tax += rate.amount_tax_max
+                rate.amount_tax_max_accum = accum_tax
```

Comment:
I think we shouldn't "compute" using the check constraints. Can you remove this line?
```suggestion
```

Also, I don't see `amount_tax_max_accum` being used anywhere. But if you want to show it, you should make the field `amount_tax_max_accum` a computed field. (it can be calculated from this line plus previous line?)